app/scripts/modules/core/src/pipeline/config/triggers/artifacts/artifact.component.ts

```diff
@@ -0,0 +1,41 @@
+import { IComponentController, IComponentOptions, module } from 'angular';
+
+import { PIPELINE_CONFIG_PROVIDER } from 'core/pipeline/config/pipelineConfigProvider';
+import { IExpectedArtifact, MissingArtifactPolicy } from 'core/domain/IExpectedArtifact';
+import { IPipeline } from 'core/domain/IPipeline';
+
+class ArtifactController implements IComponentController {
+  public artifact: IExpectedArtifact;
+  public pipeline: IPipeline;
+  public missingPolicies: string[] = Object.keys(MissingArtifactPolicy).map(key => MissingArtifactPolicy[key as any]);
+
+  public removeArtifact(): void {
+    const artifactIndex = this.pipeline.expectedArtifacts.indexOf(this.artifact);
+    this.pipeline.expectedArtifacts.splice(artifactIndex, 1);
+
+    this.pipeline.triggers.map(t => {
+      const found = t.expectedArtifacts.find(a => {
+        return this.isEqual(a, this.artifact);
+      });
+      if (found) {
+        t.expectedArtifacts = t.expectedArtifacts.filter(a => !this.isEqual(a, found));
+      }
+    });
```

Comment:
You don't need `map` since you're not doing anything with it, so should probably use `forEach` (just to be more clear that you're just iterating through the items).

Seems like this could be simplified quite a bit (unless I'm missing something):
```javascript
this.pipeline.triggers
  .forEach(t => t.expectedArtifacts = t.expectedArtifacts.filter(a => !this.isEqual(a, this.artifact)));
```

Reply:
Yep good call -- there was a conundrum with `isEqual` causing weird behavior since angular was add a `hashCode` field. I had to do some contortions to figure out what was going on and didn't revert it.